### PR TITLE
kubeadm: Improve testcases for `init` and `join` command

### PR DESCRIPTION
**Pull request by chendave**
- `cri-socket` is not allowed for mixed configuration
Set the `cri-socket` both in flags and config file will hit errors, this should not be a valid case to validate in current testcases.
- use the right methods for logging if no args are passing
- stop using of CRI endpoints without URL scheme
run the testcase with `-v` flag will reveal the warning, e.g. `W1103 ... Usage of CRI endpoints without URL scheme is deprecated...`
- improve test coverage by validating the data structure
data structure is what returned if everything okay, but this structure is not validated at all both in `init` and `join` cmd.

#### What type of PR is this?

/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by chendave** [comment]
Hey @neolit123  wdyt about this?

File: cmd/kubeadm/app/cmd/init_test.go [hunk 1]
```diff
@@ -45,7 +51,7 @@ nodeRegistration:
 apiVersion: %[1]s
 kind: ClusterConfiguration
 controlPlaneEndpoint: "3.4.5.6"
-`, kubeadmapiv1.SchemeGroupVersion.String())
+`, v1beta3.SchemeGroupVersion.String())
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init_test.go:54]: kubeadmspiv1 is better here as import name, so that we have smaller diff when the api becomes v4

**Review by neolit123 — COMMENTED** [review]
the change looks ok to me. would appreciate review from someone else too.

we can get back to this after code freeze.

**Comment by chendave** [comment]
rebased and fixed some nits.

@neolit123 @SataQiu @pacoxu  any more comments on this?

File: cmd/kubeadm/app/cmd/join_test.go [hunk 2]
```diff
@@ -206,17 +213,43 @@ func TestNewJoinData(t *testing.T) {
...
+						},
+						TLSBootstrapToken: "abcdef.0123456789abcdef",
+						Timeout:           &metav1.Duration{Duration: kubeadmapiv1.DefaultDiscoveryTimeout},
+					},
+					ControlPlane: &kubeadmapi.JoinControlPlane{
+						CertificateKey: "c39a18bae4a72e71b178661f437363da218a3efb83ddb03f1cd91d9ae1da41bd",
+					},
+				},
+				ignorePreflightErrors: sets.New("c", "d"),
+			},
```
Comment (pacoxu) [inline cmd/kubeadm/app/cmd/join_test.go:241]: I would prefer to change it to something like below.
```
			name: "--node-name flags override config from file",
			flags: map[string]string{
				options.CfgPath:  configFilePath,
				options.NodeName: "joinName",
			},
			validate: func(t *testing.T, data *joinData) {
				// validate that node-name is overwritten
				if data.cfg.NodeRegistration.Name != "joinName" {
					t.Error("Invalid NodeRegistration.Name")
				}
				if diff := cmp.Diff(joinData1, data, cmp.AllowUnexported(joinData{}), cmpopts.IgnoreFields(joinData{}, "client", "initCfg", "cfg.ControlPlane.LocalAPIEndpoint")); diff != "" {
					t.Fatalf("newJoinData returned data (-want,+got):\n%s", diff)
				}
			},
```

`validate    func(*testing.T, *joinData)`  can do the https://github.com/kubernetes/kubernetes/pull/113574/files#diff-1966268cd39e61d4b28b4c316f2073c64cf2337ae62e0e75ebb44c3d9abcf9f5R343-R347 check.

File: cmd/kubeadm/app/cmd/join_test.go [hunk 3]
```diff
@@ -42,12 +47,12 @@ discovery:
...
   certificateKey: c39a18bae4a72e71b178661f437363da218a3efb83ddb03f1cd91d9ae1da41bd
 nodeRegistration:
-  criSocket: /run/containerd/containerd.sock
+  criSocket: unix:///var/run/containerd/containerd.sock
   name: someName
   ignorePreflightErrors:
     - c
     - d
-`, kubeadmapiv1.SchemeGroupVersion.String())
+`, v1beta3.SchemeGroupVersion.String())
```
[the remark this answers is not in the document]
Comment (chendave) [reply cmd/kubeadm/app/cmd/join_test.go]: I was hoping to remove the alias actually, this has been reverted already:  https://github.com/kubernetes/kubernetes/pull/113574#discussion_r1030036834

File: cmd/kubeadm/app/cmd/join_test.go (hunk 2 above)
Comment (chendave) [reply cmd/kubeadm/app/cmd/join_test.go:249]: Done

**Comment by chendave** [comment]
@pacoxu thanks for the review, here is the diff, PTAL. https://github.com/kubernetes/kubernetes/compare/31e601eda03d857953dbb9c1d16ae088d8f81f32..32b2de1eec8f1219b35b577681089454dd86edf3

File: cmd/kubeadm/app/cmd/init_test.go [hunk 4]
```diff
@@ -98,17 +104,54 @@ func TestNewInitData(t *testing.T) {
...
+								},
+								Groups: []string{"system:bootstrappers:kubeadm:default-node-token"},
+							},
+						},
+					},
+				}
+				// validate that node-name is overwritten
+				if data.cfg.NodeRegistration.Name != "somename" {
+					t.Error("Invalid NodeRegistration.Name")
+				}
```
Comment (chendave) [inline cmd/kubeadm/app/cmd/init_test.go:139]: now, this is not needed anymore

File: cmd/kubeadm/app/cmd/join_test.go [hunk 5]
```diff
@@ -206,17 +212,52 @@ func TestNewJoinData(t *testing.T) {
...
+						ControlPlane: &kubeadmapi.JoinControlPlane{
+							CertificateKey: "c39a18bae4a72e71b178661f437363da218a3efb83ddb03f1cd91d9ae1da41bd",
+						},
+					},
+					ignorePreflightErrors: sets.New("c", "d"),
+				}
+				// validate that node-name is overwritten
+				if data.cfg.NodeRegistration.Name != "somename" {
+					t.Error("Invalid NodeRegistration.Name")
+				}
```
Comment (chendave) [inline cmd/kubeadm/app/cmd/join_test.go:245]: now, this is not needed anymore

**Review by SataQiu — COMMENTED** [review]
/lgtm

**Comment by chendave** [comment]
/assign @neolit123 

this needs someone to approve :)

**Review by SataQiu — APPROVED** [review]

**Comment by SataQiu** [comment]
/approve
thanks

**Pull request merged**